### PR TITLE
Send reset password [API], update README, update C.R.U.D. in tests

### PR DESCRIPTION
**Pull request by greenc123**

### CODE REVIEW

File: src/routes/router.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const express = require('express')
```
Comment (kimong1) [inline src/routes/router.js:1]: We don't need this file

File: src/routes/userRoutes.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    where: {
+      email: email
+    }
+  })
+  if (!userAccount) {
+    return res.status(400).json({error: {message: "account does not exist"}})
+  }
+  logger.info(`user account found for user ${userAccount.id}, sending email now`)
+  try {
+    sendPasswordResetEmail(userAccount)
```
Comment (songz) [inline src/routes/userRoutes.js:24]: await

File: src/routes/userRoutes.test.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  test('should throw error if req body does not have email', async () => {
+    const req = {
+      body: {}
+    }
+
+    await resetPassword(req, res)
+    expect(res.status.mock.calls[0][0]).toEqual(400)
+    expect(mockJson.mock.calls[0][0]).toEqual({error: {message: "invalid input"}})
+  })
+  test('should throw error if user account is not found', async () => {
```
Comment (songz) [inline src/routes/userRoutes.test.js:29]: wrong description. Should send response with status 400 if user account does not exist

File: src/routes/userRoutes.test.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+   const req = {
+     body: {
+       email: 'hello@world.com'
+     } 
+   } 
+
+    await resetPassword(req, res)
+    expect(res.status.mock.calls[0][0]).toEqual(400)
+    expect(mockJson.mock.calls[0][0]).toEqual({error: {message: "account does not exist"}})
+  })
```
Comment (songz) [inline src/routes/userRoutes.test.js:45]: missing 2 more test cases:

1. should send success if send password is successful
2. should send 500 error if send password throws error

File: src/routes/userRoutes.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    }
+  })
+  if (!userAccount) {
+    return res.status(400).json({error: {message: "account does not exist"}})
+  }
+  logger.info(`user account found for user ${userAccount.id}, sending email now`)
+  try {
+    await sendPasswordResetEmail(userAccount) 
+    logger.info(`user resent password email sent to user ${userAccount.id}`)
+    return res.status(200).send('success')
```
Comment (songz) [inline src/routes/userRoutes.js:26]: be consistent. If you are sending back `json` for all the other responses, you should send back a response here too. Instead, you are sending back a string.

File: src/routes/userRoutes.test.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    const req = {
+      body: {
+        email: 'hello@world.com' 
+      } 
+    }
+
+    await resetPassword(req, res)
+    expect(res.status.mock.calls[0][0]).toEqual(200)
+    expect(mockJson.mock.calls[0][0]).toEqual({error: {message: 'email sucessfully sent'}})
```
Comment (kimong1) [inline src/routes/userRoutes.test.js:91]: Shouldn't it be `{sucess: {message: "Email sucessfully sent"}}`?

File: src/routes/userRoutes.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  if (!userAccount) {
+    return res.status(400).json({error: {message: "Account does not exist"}})
+  }
+
+  logger.info(`User account found for user ${userAccount.id}, sending email now`)
+  try {
+    await sendPasswordResetEmail(userAccount)
+
+    logger.info(`user reset password email sent to user ${userAccount.id}`)
+    return res.status(200).send({sucess: {message: "Email sucessfully sent"}})
```
Comment (kimong1) [inline src/routes/userRoutes.js:31]: `send` -> `json`

File: src/routes/userRoutes.test.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+   const req = {
+     body: {
+       email: 'hello@world.com',
+     } 
+   } 
+
+    await resetPassword(req, res)
+    expect(res.status.mock.calls[0][0]).toEqual(400)
+    expect(mockJson.mock.calls[0][0]).toEqual({error: {message: "account does not exist"}})
```
Comment (kimong1) [inline src/routes/userRoutes.test.js:44]: image
This part failed cuz there's a typo in `account`

File: src/routes/userRoutes.test.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    })
+
+    const req = {
+      body: {
+        email: 'hello@world.com'
+      }
+    }
+
+    await resetPassword(req, res)
+    expect(res.status.mock.calls[0][0]).toEqual(500)
```
Comment (kimong1) [inline src/routes/userRoutes.test.js:65]: image
It recieved `200` which means it didn't throw error
image
1. `resetPassword` doesn't throw any error so the description is wrong. It sends json object that has error property.
2. It sends error when `email` is falsey. Which means we should put no email value in `req.body` in order to recieve 400 error

Comment (kimong1) [reply src/routes/userRoutes.test.js:65]: *In `userRoutes.js`
image
`resetPassword` sends `500` status when `sendPasswordResetEmail` throws error
*In `users.js`
image
`users.sendpasswordResetEmail` may throw error when `email.sendPasswordResetEmail` throws error.
(email is required from `mailer.js`)
*In mailer.js
image
`mgModule.sendPasswordResetEmail` will throw error when `mg.messages().send(data)` fails. `data` has `to` property whose value is `receiver` which is email address. So I assume you can make it throw error by passing a wrong email address

Comment (kimong1) [reply src/routes/userRoutes.test.js:65]: Or you can simply mock `users.sendPasswordResetEmail` then do `sendPasswordResetEmail.mockImplement(() => { throw new Error('error') })`

File: src/routes/userRoutes.test.js (hunk 6 above)
Comment (greenc123) [reply src/routes/userRoutes.test.js:91]: yes

Comment (greenc123) [reply src/routes/userRoutes.test.js:91]: good catch

**Pull request merged**